jni/tests/faiss_wrapper_unit_test.cpp

```diff
@@ -0,0 +1,251 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+#include "faiss_wrapper.h"
+
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "jni_util.h"
+#include "jni.h"
+#include "test_util.h"
+#include "faiss/IndexHNSW.h"
+#include "faiss/IndexIDMap.h"
+
+using ::testing::NiceMock;
+
+using idx_t = faiss::idx_t;
+
+struct MockIndex : faiss::IndexHNSW {
+    explicit MockIndex(idx_t d) : faiss::IndexHNSW(d, 32) {
+    }
+};
+
+
+struct MockIdMap : faiss::IndexIDMap {
+    mutable idx_t nCalled;
+    mutable const float *xCalled;
+    mutable idx_t kCalled;
+    mutable float *distancesCalled;
+    mutable idx_t *labelsCalled;
+    mutable const faiss::SearchParametersHNSW *paramsCalled;
+
+    explicit MockIdMap(MockIndex *index) : faiss::IndexIDMapTemplate<faiss::Index>(index) {
+    }
+
+    void search(
+        idx_t n,
+        const float *x,
+        idx_t k,
+        float *distances,
+        idx_t *labels,
+        const faiss::SearchParameters *params) const override {
+        nCalled = n;
+        xCalled = x;
+        kCalled = k;
+        distancesCalled = distances;
+        labelsCalled = labels;
+        paramsCalled = dynamic_cast<const faiss::SearchParametersHNSW *>(params);
+    }
+
+    void resetMock() const {
+        nCalled = 0;
+        xCalled = nullptr;
+        kCalled = 0;
+        distancesCalled = nullptr;
+        labelsCalled = nullptr;
+        paramsCalled = nullptr;
+    }
+};
+
+struct QueryIndexHNSWTestInput {
+    string description;
+    int k;
+    int efSearch;
+    int filterIdType;
+    bool filterIdsPresent;
+    bool parentIdsPresent;
+};
+
+
+
+class FaissWrappeterParametrizedTestFixture : public testing::TestWithParam<QueryIndexHNSWTestInput> {
+public:
+    FaissWrappeterParametrizedTestFixture() : index_(3), id_map_(&index_) {
+        index_.hnsw.efSearch = 100; // assigning 100 to make sure default of 16 is not used anywhere
+    };
+
+protected:
+    MockIndex index_;
+    MockIdMap id_map_;
+};
+
+namespace query_index_test {
```

Comment:
out of curiousity, what is reasoning for namespaces here?

Reply:
If I remember correctly, this allowed me to have multiple parametrized tests in the same file since I am using the same fixture for both. Could have gotten away with it if I was using different fixtures